### PR TITLE
fix: parameters when creating credentials without revocation

### PR DESCRIPTION
**Pull request by genaris**
When attempting to create some credentials without any kind of revocation configuration, under JS I've noticed that it was not properly handling some optional parameters for creating and processing credentials:
- In `create_credential`, `revocationConfiguration` was sending a default CredRevInfoStruct, that library was not interpreting as null so it was throwing an error about missing tails file path
- Also in `create_credential`, JS wrapper was not accepting an empty  `revocationStatusList`. In addition, once solved, Rust part was throwing an 'Invalid object handle' error when trying to load the object as it was mandatory
- In `process_credential`, JS wrapper was not accepting an empty `revocationRegistryDefinition`

I'm not sure if the solutions implemented in this PR are completely correct, even if they seem to work fine. So please @whalelephant and @blu3beri (as the authors of this revocation stuff in Rust and FFI wrapper in JS) let me know about any suggestion you may have (or feel free to add some commits to this PR or even completely disregard it if it's too bad! :stuck_out_tongue: )

### CODE REVIEW

**Review by TimoGlastra — APPROVED** [review]
LGTM, but I'm not sure on the optional loading in Rust / the passing of `0`,  we need the rust pros for this 🦀

File: src/ffi/credential.rs [hunk 1]
```diff
@@ -125,7 +125,11 @@ pub extern "C" fn anoncreds_create_credential(
             cred_request.load()?.cast_ref()?,
             cred_values.into(),
             rev_reg_id,
-            rev_status_list.load()?.cast_ref().ok(),
+            rev_status_list
+                .opt_load()?
+                .as_ref()
+                .map(AnonCredsObject::cast_ref)
+                .transpose()?,
```
Comment (blu3beri) [inline src/ffi/credential.rs:132]:
```suggestion
            rev_status_list
                .opt_load()?
                .cast_ref()
                .transpose()?,
```

Would this not work?

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts [hunk 2]
```diff
@@ -172,8 +172,8 @@ export class NodeJSAnoncreds implements Anoncreds {
       attributeRawValues as unknown as Buffer,
       attributeEncodedValues as unknown as Buffer,
       revocationRegistryId,
-      revocationStatusList,
-      revocationConfiguration.ref(),
+      revocationStatusList ?? 0,
```
Comment (blu3beri) [inline wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:175]: I think we just have to pass `undefined` for this and not `0`.

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts [hunk 3]
```diff
@@ -172,8 +172,8 @@ export class NodeJSAnoncreds implements Anoncreds {
...
+      revocationConfiguration ? revocationConfiguration.ref().address() : 0,
```
Comment (blu3beri) [inline wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:176]: Same for here. I think just the value / undefined works.

File: wrappers/javascript/anoncreds-nodejs/src/library/bindings.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ export const nativeBindings = {
       StringListStruct,
       FFI_STRING,
       FFI_OBJECT_HANDLE,
-      refType(CredRevInfoStruct),
+      FFI_OBJECT_HANDLE,
```
Comment (blu3beri) [inline wrappers/javascript/anoncreds-nodejs/src/library/bindings.ts:38]: GH says some values are not used in this file anymore. Can you remove them?

**Review by blu3beri — DISMISSED** [review]

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts (hunk 2 above)
Comment (genaris) [reply wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:175]: The problem is it's expecting a number, so if I send undefined I get

```TypeError: error setting argument 9 - writeUInt64: Number/String 64-bit value required```

Would it be more elegant to change the code to something like

```ts
      revocationStatusList ?? new ObjectHandle(0).handle,
```
?

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts (hunk 3 above)
Comment (genaris) [reply wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:176]: In this case it's at compilation time where I get the `number | undefined is not assignable to number`. Maybe there is a way for refType to allow it?

File: src/ffi/credential.rs (hunk 1 above)
Comment (genaris) [reply src/ffi/credential.rs:132]: If I do it directly it gives me a compilation error:

```
error[E0599]: no method named `cast_ref` found for enum `std::option::Option` in the current scope
```

**Comment by genaris** [comment]
@blu3beri just to make a bigger controversy, I've just added some fixes for optional parameters in `CredentialEntryStruct`: 

- timestamp: I had to default it to -1 because otherwise anoncreds would take it as valid
- rev_state: Defaulted to 0 as other cases in this PR

So basically the question is how can we define the structures in order to accept these values, or how to find an elegant way to define them.

File: src/ffi/credential.rs (hunk 1 above)
Comment (blu3beri) [reply src/ffi/credential.rs:132]: Okay yeah that makes sense. I think your implementation is good.

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts [hunk 5]
```diff
@@ -172,8 +171,8 @@ export class NodeJSAnoncreds implements Anoncreds {
       attributeRawValues as unknown as Buffer,
       attributeEncodedValues as unknown as Buffer,
       revocationRegistryId,
-      revocationStatusList,
-      revocationConfiguration.ref(),
+      options.revocationStatusList?.handle ?? 0,
```
Comment (blu3beri) [inline wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:174]: Isn't this handled in the `serializeArguments` function? I would strongly prefer not to send the `0` as that can be a valid handle number. 

If `serializeArguments` does not handle this we should 100% do it there. Maybe we should just use `USIZE::MAX`?

**Review by blu3beri — COMMENTED** [review]
Just one comment regarding the use of `0` for an unexacting handle type.

File: wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts (hunk 5 above)
Comment (genaris) [reply wrappers/javascript/anoncreds-nodejs/src/NodeJSAnoncreds.ts:174]: The problem with `serializeArguments` is that it receives the object as undefined so it does not know which type of object it is supposed to be (i.e. typeof arg = 'undefined'). So it returns NULL, which works fine for optional strings but not for optional object handles (whose type is `size_t`).

**Review by TimoGlastra — APPROVED** [review]

**Pull request merged**